python/ray/experimental/workflow/serialization.py

```diff
@@ -0,0 +1,114 @@
+            common.STORAGE_ACTOR_NAME,
+            namespace=common.MANAGEMENT_ACTOR_NAMESPACE)
+    except ValueError:
+        store = storage.get_global_storage()
+        if warn_on_creation:
+            logger.warning("Cannot access workflow serialization manager. It "
+                           "could be because "
+                           "the workflow manager exited unexpectedly. A new "
+                           "workflow manager is being created with storage "
+                           f"'{store}'.")
+        handle = Manager.options(
+            name=common.STORAGE_ACTOR_NAME,
+            namespace=common.MANAGEMENT_ACTOR_NAMESPACE,
+            lifetime="detached").remote(store)
+        ray.get(handle.ping.remote())
+        return handle
+
+
+@dataclass
+class Upload:
+    identifier_ref: ObjectRef[str]
+    upload_task: ObjectRef[None]
+
+
+@ray.remote
+def _put_helper(identifier: str, obj: Any,
+                wf_storage: "workflow_storage.WorkflowStorage") -> None:
+    if isinstance(obj, ray.ObjectRef):
+        raise NotImplementedError("Workflow does not support checkpointing "
+                                  "nested object references yet.")
+    paths = wf_storage._key_obj_id(identifier)
+    asyncio.get_event_loop().run_until_complete(wf_storage._put(paths, obj))
+
+
+@ray.remote(num_cpus=0)
+class Manager:
+    """
+    Responsible for deduping the serialization/upload of object references.
+    """
+
+    def __init__(self, storage: storage.Storage):
+        self._uploads: Dict[ray.ObjectRef, Upload] = {}
+        self._storage = storage
+        self._num_uploads = 0
+
+    def ping(self) -> None:
+        """
+        Trivial function to ensure actor creation is successful.
+        """
+        return None
+
+    async def save_objectref(
+            self, ref_tuple: Tuple[ray.ObjectRef],
+            workflow_id: "str") -> Tuple[List[str], ray.ObjectRef]:
+        """Serialize and upload an object reference exactly once.
+
+        Args:
+            ref_tuple: A 1-element tuple which wraps the reference.
+
+        Returns:
+            A pair. The first element is the paths the ref will be uploaded to.
+            The second is an object reference to the upload task.
+        """
+        wf_storage = workflow_storage.WorkflowStorage(workflow_id,
+                                                      self._storage)
+        ref, = ref_tuple
+        # Use the hex as the key to avoid holding a reference to the object.
+        key = ref.hex()
+
+        if key not in self._uploads:
+            # TODO(Alex): We should probably eventually free these refs.
+            identifier_ref = common.calculate_identifier.remote(ref)
+            upload_task = _put_helper.remote(identifier_ref, ref, wf_storage)
+            self._uploads[key] = Upload(identifier_ref, upload_task)
+            self._num_uploads += 1
+
+        info = self._uploads[key]
+        identifer = await info.identifier_ref
```

Comment:
Not quite sure how async actor here works. Will the await be done on the client-side or in the remote? Does this mean at a given time, the remote will have several coroutines which are awaiting something?

See here it awaits anyway, should we just store the deref-identifier there?

Reply:
hmmm i'm not quite sure i understand what you mean by client side and remote here, but I think the way that `await` works is it yields to the asyncio scheduler, then does `ray.get`. So the calculate_identifier task does the work, and the coroutine sleeps until the task is finished.